### PR TITLE
Initial Version

### PR DESCRIPTION
**Pull request by rogerluan**
This is just the first version. What's missing (and will be added in separate PRs):

- README
- CI
- Code coverage report (it's at 100% code coverage currently)
- Might change how configs are managed

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+Style/RedundantFetchBlock:
+  Enabled: true
+
+# Use $stdout instead of STDOUT.
+Style/GlobalStdStream:
+  Enabled: true
+
+# Prefer Api::V1::MyController instead of nested and indented modules
+Style/ClassAndModuleChildren:
+  Enabled: false
```
Comment (matolah) [inline .rubocop.yml]: Shouldn't this rule only be applied to files in the `app` folder?

File: Gemfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+source "https://rubygems.org"
+
+gemspec
+
+gem "rspec", "~> 3"
+gem "rubocop-rspec", "~> 2"
+gem "pry"
+gem "dotenv"
+gem "simplecov"
```
Comment (matolah) [inline Gemfile]: I think it would be cleaner if the dependencies are declared in `hoov_vin.gemspec`. Up to you though

File: lib/vin.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+require "redis"
+require "vin/mixins/redis"
+
+class VIN
+  extend VIN::Mixins::Redis
+
+  def initialize(config: nil)
+    @config = config || VIN::Config.new
+  end
```
Comment (matolah) [inline lib/vin.rb:11]: Maybe you could create a `configure` method that yield a `config` and initialize this in a fancier way? 🪄

File: lib/vin.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      return ids unless ids.length > initial_id_count
+    end
+    ids
+  end
+
+  private
+
+  def generator
+    @generator ||= Generator.new(config: @config)
+  end
```
Comment (matolah) [inline lib/vin.rb:34]: Is `config` really a property of `VIN`? It seems like it is a property of `Generator` and we could `delegate` to `generator.config`

File: lib/vin/config.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+class VIN
+  class Config
```
Comment (matolah) [inline lib/vin/config.rb:4]: Really liked this Config class ❤️

File: lib/vin/lua_script.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+require "erb"
+require "vin/config"
+
+class VIN
+  module LuaScript
```
Comment (matolah) [inline lib/vin/lua_script.rb:7]: No idea what's going on here but it looks very good

File: lib/vin/response.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      redis_response[SECONDS_INDEX]
+    end
+
+    def microseconds_part
+      redis_response[MICROSECONDS_INDEX]
+    end
+
+    private
+
+    attr_reader :redis_response
```
Comment (matolah) [inline lib/vin/response.rb:41]: Shouldn't this be before the methods?

File: spec/lib/vin/timestamp_spec.rb [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    context "when epoch is custom" do
+      let(:milliseconds) { 7_948_800_000 }
+      let(:epoch) { 1_483_228_800_000 }
+
+      it { expect(subject.with_unix_epoch.milliseconds).to(be(1_491_177_600_000)) }
+    end
+  end
+
+  describe ".from_reids" do
```
Comment (matolah) [inline spec/lib/vin/timestamp_spec.rb]:
```suggestion
  describe ".from_redis" do
```

**Review by matolah — CHANGES_REQUESTED** [review]
Nice stuff 🚀

File: Gemfile (hunk 2 above)
Comment (rogerluan) [reply Gemfile]: Best practice is to add development dependencies to the Gemfile, and required dependencies in the gemspec 😊 there's even a rubocop rule that warns against this

File: lib/vin/response.rb (hunk 7 above)
Comment (rogerluan) [reply lib/vin/response.rb:41]: This attribute is private, so it must be below the `private` keyword here

File: .rubocop.yml (hunk 1 above)
Comment (rogerluan) [reply .rubocop.yml]: There's no app folder as this is not a rails app, it's just a gem 😅

File: lib/vin.rb (hunk 3 above)
Comment (rogerluan) [reply lib/vin.rb:11]: Great suggestion!! ❤️ 

You and I discussed this and we agreed to do this later: https://github.com/hoovbr/vin/issues/3

File: lib/vin.rb (hunk 4 above)
Comment (rogerluan) [reply lib/vin.rb:34]: The reason it was done like this is to abstract the "generator" from the user when initializing the VIN singleton from Rails. Both approaches have pros/cons but I think this was a bit cleaner 🙏

**Review by matolah — APPROVED** [review]

**Pull request merged**